File: Labs/StudentTRegression.jl
# Gibbs sampling for the regression model with student-t errors
# Author: Mattias Villani, for Lab 4 of the course Advanced Bayesian Learning

using Plots
using Distributions
using LaTeXStrings
using DelimitedFiles
using LinearAlgebra
import ColorSchemes: Paired_12; colors = Paired_12
basicColors = Paired_12[[1,2,3,4,7,8]]

# Defining some distributions
ScaledInverseChiSq(ν,τ²) = InverseGamma(ν/2,ν*τ²/2)
TDistGeneral(μ,σ,ν) = LocationScale(μ,σ,TDist(ν))

"""
    Gibbs sampling for the regression model with student-t errors
        y = x'*β + ε, ε iid ~ t_ν(0,σ,μ)
    with prior
        β | σ ~ N(0,τ^2σ^2*I), log(σ) ∼ Uniform
"""
function GibbsTReg(y, X, ν, τ, nIter)
    n, q = size(X)
    β = X \ y
    ε = y - X*β
    σ = norm(ε)/√(n-q)
    postDraws = zeros(nIter,q+1)
    for i in 1:nIter

        # Update U
        U = [rand(ScaledInverseChiSq(ν+1,(ν*σ^2 + εi^2)/(ν+1))) for εi in ε]

        # Update β
        A = Diagonal(ones(n)./(.√U))
        yt = A*y
        Xt = A*X
        βhat = Xt \ yt # Least squares estimate
        Σpost = inv(Xt'*Xt + τ^(-2)*I(q))
        μpost = vec(Σpost*Xt'*Xt*βhat)
        β = rand(MvNormal(μpost,Σpost))
        postDraws[i,1:q] = β
        ε = y - X*β

        # Update σ
        σ = √rand(Gamma(n*ν/2, 1/((ν/2)*sum(ones(n)./U))))
        postDraws[i,q+1] = σ

    end # end Gibbs iterations
    return postDraws
end # end function GibbsStudentTReg


"""
    Gibbs sampling for the regression model with student-t errors
        y = x'*β + ε, ε iid ~ t_ν(0,σ,μ)
    with prior
        β | σ ~ N(0,τ^2σ^2*I), log(σ) ∼ Uniform
"""
function GibbsTRegDF(y, X, τ, nIter)
    n, q = size(X)
    β = X \ y
    ε = y - X*β
    σ = norm(ε)/√(n-q)
    excessKurtosis = max(mean((ε/σ).^4)-3, eps())
    ν = min(4 + 6/excessKurtosis, 30)
    postDraws = zeros(nIter,q+2)
    logLike(ν, σ², u) = sum([logpdf(ScaledInverseChiSq(ν, σ²), ui) for ui ∈ u])
    for i in 1:nIter

        # Update U
        U = [rand(ScaledInverseChiSq(ν+1,(ν*σ^2 + εi^2)/(ν+1))) for εi in ε]

        # Update β
        A = Diagonal(ones(n)./(.√U))
        yt = A*y
        Xt = A*X
        βhat = Xt \ yt # Least squares estimate
        Σpost = inv(Xt'*Xt + τ^(-2)*I(q))
        μpost = vec(Σpost*Xt'*Xt*βhat)
        β = rand(MvNormal(μpost,Σpost))
        postDraws[i,1:q] = β
        ε = y - X*β

        # Update σ
        σ = √rand(Gamma(n*ν/2, 1/((ν/2)*sum(ones(n)./U))))
        postDraws[i,q+1] = σ

        # Update ν
        ψ = 0.1 # RWM scaling
        logPostCurr = logLike(ν, σ^2, U) 
        νlogProp = rand(Normal(log(ν), ψ))
        logPostProp = logLike(exp(νlogProp), σ^2, U) 
        accProb = min(1,exp(logPostProp-logPostCurr))
        if rand() < accProb
            ν = exp(νlogProp)
        else
            # Do nothing
        end
        postDraws[i,q+2] = ν

    end # end Gibbs iterations
    return postDraws
end # end function GibbsTRegDF




"""
    Predictive distribution for the regression model with student-t errors
        y = x'*β + ε, ε iid ~ t(0,σ,μ)
    with prior
        β | σ ~ N(0,τσ*I), log(σ) ∼ Uniform
"""
function PredTReg(X, postDraws, ν)
    nPreds,q = size(X)
    nSim = size(postDraws,1)
    yPreds = zeros(nSim,nPreds)
    for i = 1:nSim
        β = postDraws[i,1:q]
        σ = postDraws[i,end]
        yPreds[i,:] = [rand(TDistGeneral(μ, σ, ν)) for μ in X*β]
    end
    return yPreds
end

analyzeData = false
if analyzeData
    # Read data from file
    data = readdlm(download("https://github.com/mattiasvillani/AdvBayesLearnCourse/raw/master/Labs/regression.csv"))
    X = [ones(size(data,1),1) data[:,1]]
    y = data[:,2]
    scatter(X[:,2],y)

    # Simulating from the posterior of β and σ
    ν = 2
    τ = 1
    nIter = 10000
    postDraws = GibbsTReg(y, X, ν, τ, nIter)
    mean(postDraws, dims = 1)
    p1 = histogram(postDraws[:,1], nbins = 100, title = L"\beta_0");
    p2 = histogram(postDraws[:,2], nbins = 100, title = L"\beta_1");
    p3 = histogram(postDraws[:,3].^2*(ν/(ν-2)), nbins = 100, title = L"\mathbb{V}(y)=\sigma^2\frac{\nu}{\nu-2}");
    plot(p1,p2,p3)

    # Simulating from the predictive distribution
    xGrid = -1:0.1:1
    XPred = [ones(length(xGrid),1) xGrid]
    yPreds = PredTReg(XPred, postDraws, ν)
    predBands = zeros(size(XPred,1),5)
    for i in 1:size(yPreds,2)
        predBands[i,:] = quantile(yPreds[:,i], [0.005 0.025 0.5 0.975 0.995])
    end
    p = scatter(X[:,2],y, xlabel = L"x", ylabel = L"y");
    plot!(xGrid,predBands[:,3], color = basicColors[6])
    plot!(xGrid,predBands[:,1], color = basicColors[1])
    plot!(xGrid,predBands[:,2], color = basicColors[2])
    plot!(xGrid,predBands[:,4], color = basicColors[2])
    plot!(xGrid,predBands[:,5], color = basicColors[1])
end

# Estimating the degrees of freedom
# Read data from file
nIter = 10000
data = readdlm(download("https://github.com/mattiasvillani/AdvBayesLearnCourse/raw/master/Labs/regression.csv"))
X = [ones(size(data,1),1) data[:,1]]
y = data[:,2]
τ = 10
postDraws = GibbsTRegDF(y, X, τ, nIter)

p1 = plot(size=(900,500), postDraws[:,4], lw = 1)  
p2 = histogram(postDraws[:,4], lw = 0)
l = @layout([a b])
plot(size=(900,500), p1, p2, layout = l)
savefig(figFolder*"StudentTMCMC.png")

File: Labs/StudentTRegression.R
# Gibbs sampling for regression with student t errors
# Author: Mattias Villani, for the course Advanced Bayesian Learning

#install.packages("mvtnorm")
library(mvtnorm)

# Random number generator for scaled inverse Chi-square distribution
rScaledInvChi2 <- function(n, df, scale){
  return((df*scale)/rchisq(n,df=df))
}

# Gibbs sampler for student-t regression
GibbsTReg <- function(y, X, nu, tau, nIter){
  n = dim(X)[1]
  q = dim(X)[2]
  beta_ = solve(crossprod(X),crossprod(X,y))
  epsilon = y - X%*%beta_
  sigma = sqrt(sum(epsilon^2)/(n-q))
  postDraws = matrix(0,nIter,q+1)
  for (i in seq(1,nIter)){
    
    # Update U
    U = rScaledInvChi2(n, nu + 1, (nu*sigma^2 + epsilon^2)/(nu+1))

    # Update beta
    A = diag(as.vector(1/(sqrt(U))))
    yt = A%*%y
    Xt = A%*%X
    betaHat = solve(crossprod(Xt),crossprod(Xt,yt)) # Least squares estimate
    SigmaPost = solve(crossprod(Xt) + tau^(-2)*diag(q))
    muPost = SigmaPost%*%(crossprod(Xt)%*%betaHat)
    beta_ = t(rmvnorm(1, muPost, SigmaPost))
    postDraws[i,1:q] = beta_
    epsilon = y - X%*%beta_
    
    # Update sigma
    sigma = sqrt(rgamma(1,n*nu/2, rate = (nu/2)*sum(1/U) ) )
    postDraws[i,q+1] = sigma
    
  }
  return(postDraws)
}

# Simulating from the predictive distribution of the student-t regression
PredTReg <- function(XPred, postDraws, nu){
  nPreds = dim(XPred)[1]
  q = dim(XPred)[2]
  nSim = dim(postDraws)[1]
  yPreds = matrix(0, nSim, nPreds)
  for (i in seq(1,nSim)){
    beta_ = postDraws[i,1:q]
    sigma = postDraws[i,q+1]
    yPreds[i,] = XPred%*%beta_ + sigma*rt(nPreds, nu)
  }
  return(yPreds)
}

# Reading data from the course repo
data = read.csv("https://github.com/mattiasvillani/AdvBayesLearnCourse/raw/master/Labs/regression.csv", sep ='\t', header = FALSE)
names(data) <- c("x","y")
X = cbind(1,data[,1])
y = data[,2]
n = length(y)
plot(X[,2],y)

# Simulating from the posterior of beta and sigma
tau = 1
nu = 2
nIter = 10000
postDraws = GibbsTReg(y, X, nu, tau, nIter)
colMeans(postDraws)
apply(postDraws, 2, sd)
par(mfrow = c(2,2))
hist(postDraws[,1], 50, main = "beta0")
hist(postDraws[,2], 50, main = "beta1")
hist(postDraws[,3], 50, main = "sigma")
hist(postDraws[,3]^2*(nu/(nu-2)), 50, main = "Var(y)")

# Normal approx of posterior by optimization
logLike <- function(theta, y, X, nu){
  n = dim(X)[1]
  p = dim(X)[2]
  beta = theta[1:p]
  sigma = exp(theta[p+1])
  epsilon = (y-X%*%beta)/sigma
  return(sum(log(1/sigma) + dt(epsilon, nu, log = TRUE)))
}

logPrior <- function(theta, tau){
  p = length(theta) - 1
  beta = theta[1:p]
  sigma = exp(theta[p+1])
  logPriorPDF = dmvnorm(beta, mean = rep(0, p), sigma = tau^2*diag(p), log = TRUE) + 
    log(1/sigma)
  return(logPriorPDF)
}

logPost <- function(theta, y, X, nu, tau){
  return(logLike(theta, y, X, nu) + logPrior(theta, tau) )
}

beta_ml = solve(crossprod(X))%*%crossprod(X,y)
sigma_ml = sqrt(crossprod(y-X%*%beta_ml)/n)
theta_ml = c(beta_ml,log(sigma_ml))
logPost(theta_ml, y, X, nu, tau)
logLike(theta_ml, y, X, nu)
logPrior(theta_ml, tau)

initVal = theta_ml
OptimResults <- optim(initVal, logPost, gr=NULL, y, X, nu, tau, method=c("BFGS"), 
  control=list(fnscale=-1), hessian=TRUE)
postMode = OptimResults$par
approxPostStd <- sqrt(diag(-solve(OptimResults$hessian)))
apply(postDraws, 2, sd)

logPost(theta_ml, y, X, nu, tau)
logPost(postMode, y, X, nu, tau)
postMode


# Making predictions and plotting predictive distribution
xGrid = seq(-1,1,by =.1)
XPred = cbind(1,xGrid)
yPreds = PredTReg(XPred, postDraws, nu)
predBands = matrix(0,dim(XPred)[1],5)
for (i in 1:length(xGrid)){
  predBands[i,] = quantile(yPreds[,i], c(0.005,0.025,0.5,0.975,0.995))
}
par(mfrow = c(1,1))
plot(X[,2], y, xlab = "x", ylab = "y")
lines(xGrid,predBands[,3], col = "red")
lines(xGrid,predBands[,1], col = "green")
lines(xGrid,predBands[,2], col = "blue")
lines(xGrid,predBands[,4], col = "blue")
lines(xGrid,predBands[,5], col = "green")

